// drizzle-builder/dist/parse/parsers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _jsYaml = require('js-yaml');

var yaml = _interopRequireDefault(_jsYaml).default;

var _marked = require('marked');

var marked = _interopRequireDefault(_marked).default;

var _frontMatter = require('front-matter');

var frontMatter = _interopRequireDefault(_frontMatter).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var parsers = {
  content: {
    pattern: '\.(html|hbs|handlebars)$',
    parseFn: function parseFn(contents, path) {
      var matter = frontMatter(contents);
      return {
        contents: matter.body,
        data: matter.attributes
      };
    }
  },
  markdown: {
    pattern: '\.(md|markdown)$',
    parseFn: function parseFn(contents, path) {
      var matter = frontMatter(contents);
      return {
        contents: marked(matter.body),
        data: matter.attributes
      };
    }
  },
  yaml: {
    pattern: '\.(yaml|yml)$',
    parseFn: function parseFn(contents, path) {
      return { contents: yaml.safeLoad(contents) };
    }
  },
  json: {
    pattern: '\.json$',
    parseFn: function parseFn(contents, path) {
      return { contents: JSON.parse(contents) };
    }
  },
  default: {
    parseFn: function parseFn(contents, path) {
      return { contents: contents };
    }
  }
}; /**
    * Default parsers, merged into options. Can be overridden by user options.
    */

exports.default = parsers;
module.exports = exports['default'];
