// drizzle-builder/dist/parse/tree.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
/**
 * Build an object representing the upcoming output tree
 * @module parse/tree
 */

/**
return {
  data      : allData[0],
  pages     : allData[1],
  patterns  : allData[2],
  templates : allData[3],
  options   : options
};
**/

function walkResources(resources, options, resourceType) {
  var resourceTree = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [];

  for (var resourceKey in resources) {
    var item = resources[resourceKey];
    if (item.resourceType && item.resourceType === resourceType.singular) {
      resourceTree.push(item);
    } else {
      walkResources(item, options, resourceType, resourceTree);
    }
  }
  return resourceTree;
}

function parseTree(allData, options) {
  var dataObj = {
    data: allData[0],
    pages: allData[1],
    patterns: allData[2],
    templates: allData[3],
    tree: {
      pages: walkResources(allData[1], options, options.keys.pages),
      collections: walkResources(allData[2], options, options.keys.collections)
    },
    options: options
  };
  return dataObj;
}

exports.default = parseTree;
module.exports = exports['default'];
