// drizzle-builder/dist/render/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _pages = require('./pages');

var renderPages = _interopRequireDefault(_pages).default;

var _collections = require('./collections');

var renderCollections = _interopRequireDefault(_collections).default;

var _error = require('../utils/error');

var DrizzleError = _interopRequireDefault(_error).default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Render pages and pattern-collection pages.
 *
 * @param {Object} drizzleData  All data built so far
 * @return {Promise} resolving to drizzleData
 */
function render(drizzleData) {
  return Promise.all([renderPages(drizzleData), renderCollections(drizzleData)]).then(function (allData) {
    return {
      data: drizzleData.data,
      pages: allData[0],
      patterns: allData[1],
      templates: drizzleData.templates,
      options: drizzleData.options,
      tree: drizzleData.tree
    };
  }, function (error) {
    return DrizzleError.error(error, drizzleData.options.debug);
  });
}

exports.default = render;
module.exports = exports['default'];
